src/ContainerResolver.php

```diff
@@ -10,76 +10,108 @@
 
 namespace Laminas\Cli;
 
+use InvalidArgumentException;
 use Laminas\ModuleManager\ModuleManagerInterface;
 use Laminas\Mvc\Service\ServiceManagerConfig;
 use Laminas\ServiceManager\ServiceManager;
 use Laminas\Stdlib\ArrayUtils;
 use Psr\Container\ContainerInterface;
 use RuntimeException;
+use Symfony\Component\Console\Input\InputInterface;
 use Webmozart\Assert\Assert;
 
+use function assert;
 use function class_exists;
 use function file_exists;
+use function is_string;
+use function sprintf;
 
 /**
  * @internal
  */
 final class ContainerResolver
 {
+    /**
+     * @var string
+     * @psalm-var non-empty-string
+     */
+    private $projectRoot;
+
+    /**
+     * @psalm-param non-empty-string $projectRoot
+     */
+    public function __construct(string $projectRoot)
+    {
+        $this->projectRoot = $projectRoot;
+    }
+
     /**
      * Try to find container in Laminas application.
      * Supports out of the box Laminas MVC and Mezzio applications.
      *
      * @throws RuntimeException When cannot locate PSR-11 container for the application.
      */
-    public static function resolve(): ContainerInterface
-    {
-        if (file_exists('config/container.php')) {
-            return self::resolveDefaultContainer();
+    public function resolve(
+        InputInterface $input
+    ): ContainerInterface {
+        $pathToContainer = $input->getOption(ApplicationFactory::CONTAINER_OPTION) ?? '';
+        assert(is_string($pathToContainer));
```

Comment:
I use `assert`/`Assert` depending on what I want achieve.

--- 

When the result is always `true` and I only have to provide informations to psalm, I can either use `@var` annotation or I do an assertion which is mostly disabled in production systems.
The benefit in doing so is, that it has no runtime impact here and if something would change in the `InputInterface` implementation, it will directly fail in unit tests (which have `zend.assertion=1`).

---

In case the value I want to assert on is unknown to the LoC due to external dependency (e.g. integer which has to be positive but the annotation does not require that), I use `Assert::positiveInteger`.

---

In this case, we are defining the `container` option in the project. So there is most-likely **no** reason why to waste runtime assertion on that variable unless someone decides to somehow force the `InputInterface` implementation to not return string.

Reply:
The best example here is something like this:

```php
$pathToContainer = sprintf('%s/%s', $this->projectRoot, $pathToContainer);
```

Psalm asserts that the `$pathToContainer` is just a string, even tho, it is absolutely not empty due to the fact that there is a `/` in there **and** that `projectRoot` property is non-empty-string.

So overall, when using `assert()`, that is just to provide proper details to psalm while using `Assert` would provide the same information to psalm but has runtime impact in upstream projects in production envs.

Reply:
I think the runtime impact argument here is weak - this is the CLI, and the handful of assertions we'd be doing here are not likely to have an impact on the performance; the work the commands do themselves will be far more of an issue. I'd argue for consistency over performance in this context.

As an added bonus, webmozart/assert allows us to provide messages for failed assertions, which will then bubble up to the user.

Reply:
Okay then, I use `assert` pretty much during my daily job to avoid useless runtime assertions but I can change that here.

--- 

So do we want to clarify how to handle this in future changes in other packages? I use `assert` a lot in cases where it should not be necessary to do any assertion (I gave the example with that `sprints` LoC). If there is a coding guideline for laminas, we should add `assert` to forbidden functions in the codingstandard then?

Reply:
Something to bring up at the TSC meeting, I think, so we can then document it in the maintainers/contributor's guide(s).